polynote-kernel/src/main/scala/polynote/kernel/dependency/CoursierFetcher.scala

```diff
@@ -24,50 +24,67 @@ import coursier.error.ResolutionError
 import scala.tools.nsc.interpreter.InputStream
+import coursier.credentials.{Credentials => CoursierCredentials, DirectCredentials}
+import coursier.core.Authentication
 
 object CoursierFetcher {
   type ArtifactTask[A] = RIO[CurrentTask, A]
   type OuterTask[A] = RIO[TaskManager with CurrentTask, A]
   //type ArtifactTask[A] = RIO[CurrentTask, A]
 
   private val excludedOrgs = Set(Organization("org.scala-lang"), Organization("org.apache.spark"))
-  private val cache = FileCache[ArtifactTask]()
+  private val baseCache = FileCache[ArtifactTask]()
 
-  def fetch(language: String): RIO[CurrentNotebook with TaskManager with Blocking, List[(Boolean, String, File)]] = TaskManager.run("Coursier", "Dependencies", "Resolving dependencies") {
+  def fetch(language: String): RIO[Config with CurrentNotebook with TaskManager with Blocking, List[(Boolean, String, File)]] = TaskManager.run("Coursier", "Dependencies", "Resolving dependencies") {
     for {
-      config       <- CurrentNotebook.config
-      dependencies  = config.dependencies.flatMap(_.toMap.get(language)).map(_.toList).getOrElse(Nil)
-      (deps, uris)  = splitDependencies(dependencies)
-      repoConfigs   = config.repositories.map(_.toList).getOrElse(Nil)
-      exclusions    = config.exclusions.map(_.toList).getOrElse(Nil)
-      repositories <- ZIO.fromEither(repositories(repoConfigs))
-      resolution   <- resolution(deps, exclusions, repositories)
-      _            <- CurrentTask.update(_.copy(detail = "Downloading dependencies...", progress = 0))
-      downloadDeps <- download(resolution).fork
-      downloadUris <- downloadUris(uris).fork
-      downloaded   <- downloadDeps.join.map2(downloadUris.join)(_ ++ _)
+      polynoteConfig <- Config.access
+      config         <- CurrentNotebook.config
+      dependencies    = config.dependencies.flatMap(_.toMap.get(language)).map(_.toList).getOrElse(Nil)
+      (deps, uris)    = splitDependencies(dependencies)
+      repoConfigs     = config.repositories.map(_.toList).getOrElse(Nil)
+      exclusions      = config.exclusions.map(_.toList).getOrElse(Nil)
+      credentials    <- loadCredentials(polynoteConfig.credentials)
+      repositories   <- ZIO.fromEither(repositories(repoConfigs, credentials))
+      cache           = baseCache.addCredentials(credentials: _*)
+      resolution     <- resolution(deps, exclusions, repositories, cache)
+      _              <- CurrentTask.update(_.copy(detail = "Downloading dependencies...", progress = 0))
+      downloadDeps   <- download(resolution, cache).fork
+      downloadUris   <- downloadUris(uris).fork
+      downloaded     <- downloadDeps.join.map2(downloadUris.join)(_ ++ _)
     } yield downloaded
   }
 
-  private def repositories(repositories: List[RepositoryConfig]): Either[Throwable, List[Repository]] = repositories.collect {
+  private def loadCredentials(credentials: CredentialsConfig): Task[List[DirectCredentials]] = credentials.coursier match {
```

Comment:
Hmmm, now we have to think about what should happen if they specified an invalid path, or there was a problem parsing the credentials, etc.

My instinct is that we should `catchAll` in the `CoursierCredentials(...).get().toList` task, log the error, and return `Nil`. Polynote will still work, but credentials won't be loaded. Then `loadCredentials` could return UIO as well.

Reply:
Otherwise this looks good!

Reply:
Changed in 692ac68 (Unless you meant other kind of logging)